### ISSUE TITLE
Refactor: preprocessing and dimensionality reduction

### ISSUE THREAD

**Issue by ContiPaolo** [alan-turing-institute/autoemulate#348]
Dimensionality reduction (for both inputs and outputs) is essential for handling large-scale spatio-temporal data. This requires training and, if using deep learning, hyper-parameter (HP) tuning.

We need to define how to handle this in the refactoring process. There are three possible approaches:

1. **Embedded in the emulator**: Dimensionality reduction and emulator training happen together. More general but computationally expensive.
2. **Separate pre-step:** Train dimensionality reduction first, then train the emulator on reduced data. More efficient but might lead to a suboptimal combination of reduction + emulator.
3. **Hybrid approach (recommended):** First, train dimensionality reduction and compare emulator performance. Then, retrain both together to fine-tune the system just on the best selected model.

In terms of refactoring, we might choose an approach (maybe the hybrid) as default, but I think we should flexibility to switch between methods. This might require making the preprocessing step a trainable component that can be:
- Trained together with the emulator (approach 1),
- Trained separately beforehand (approach 2),
- Retrained after emulator selection (approach 3).

**Comment by sgreenbury** [comment]
Thanks @ContiPaolo for adding this, this is very helpful and I like the approaches outlined! Relates to #323 too - it seems like the options might be also generally applicable to any processing that is trainable (as opposed to some that might be single transformations of the data, e.g. z-score).

For approach 1/3, I'm wondering how we flexibly integrate different dimensionality reduction/trainable processing approaches with emulators. Perhaps creating `nn.Sequential` model with an emulator and processor that both implement the PyTorch backend so that they can be fit with the same API could work. A custom loss might also be required in implementing to also optimise the processing aspect. Being able to compose different processing models and emulators at runtime could be a nice feature too to allow searching over different model compositions.

**Comment by radka-j** [comment]
We agreed that we will handle distribution transforms as a separate issue. There already is one with notes on potential approaches in #376.

**Comment by sgreenbury** [comment]
Thanks @radka-j for connecting to #376, adding some ideas for the specific tasks to cover here for this issue (and we might consider revising some of these further into sub-issues too):
- [ ] Add AutoEmulate `Transform` class
- [ ] Add AutoEmulate `TransformModule` (also an `nn.Module`) (this could be a separate issue)
- [ ] Implement the transform for PCA and VAE in current main
- [ ] Add classes for emulator
  - [ ] With fixed preprocessing and postprocessing
  - [ ] Learnable pre/post-processing (this could be a separate issue)
- [ ] Implement initial approach (e.g. sampling #376) for handling distribution output. This would enable initial functionality with emulators that return distributions.
- [ ] Consider API design wrt approaches that might be included in #376
- [ ] Apply the new emulator(s) with transforms to examples (e.g. https://github.com/alan-turing-institute/autoemulate/blob/main/docs/tutorials/05_reaction_diffusion.ipynb)

**Comment by sgreenbury** [comment]
Following discussion with @ContiPaolo:
- Transforms would benefit from refit method with option to retrain the transforms on final data (e.g. all the data not just train split)
- Adding a method to inverse transform for estimating distributional properties (like std_dev, covariance). This enables an empirical estimate of properties of transformed distributions
- For transformations of covariance with analytical form (like for PCA) we can compare with the sampling approach for validating
- Current API should be flexible for non-linear methods (e.g. delta method) in #376
